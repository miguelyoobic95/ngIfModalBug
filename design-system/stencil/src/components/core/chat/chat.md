---
name: Chat
category: Core Components
---
The Chat component is a component designed to display conversation between two or more people.

## Message Interface

``` typescript
    export interface IChatMessage {
        content: string;
        author?: string;
        time?: string;
        isAlternate?: boolean;
        img?: string;
    }
```
|Name|Description|
|---|---|
|`content`|Content of the message|
|`author`|Author of the message|
|`time`|Time of the message|
|`isAlternate`|set to true if the message has been send by the current user|
|`img`|Source of the image displayed in the Chat|

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`heading`|string|   |Heading of the Chat|
|`messages`|IChatMessage[]|null|Messages of the Chat|

```yoo-chat.html
    <yoo-chat heading="Chat Title"></yoo-chat>
```

```yoo-chat.js
    var chat = document.querySelector('yoo-chat');
    chat.messages = [{content: "First message",author:"Seb",time:"2/05/18 8:47am"},
      {content: "2nd message: It seems to work fine but I'm sure there is a bug somewhere !",author:"Seb",
      img: "https://res.cloudinary.com/www-yoobic-com/image/upload/w_1020,h_480,c_fill,q_auto:low,f_auto,fl_lossy/a_exif/v1522152485/k7dedr96ainrdhmkzlgj.png", time:"2/05/18 8:51am"},
      {content: "Third message, it is a very very very very long message. Not so long after all !",author:"Mike",isAlternate: true, time:"2/05/18 9:34am"}];
```

## CSS

|Type|Name|Description|
|---|---|---|

## Events
|Attr|Description|
|---|---|---|
