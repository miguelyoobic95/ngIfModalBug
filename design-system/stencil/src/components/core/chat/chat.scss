

:host{
    height: 100%;
    .outer-container{
        height: 100%;
        background-color: cssvar(light); 
        .chat-header{
            text-align: center;
            //border-bottom: 1px solid cssvar(dark-20);
        }
        .load-more {
            justify-content:center;
            background: cssvar(dark-10);
            color: cssvar(text-color);
        }
        .scroll-container {
            height: 100%;
        }
        .messages-container{
            // padding: 1rem;
            padding-right: 1rem;
            padding-left: 1rem;
            margin-top: 0.2rem;
            // margin-bottom: 0.2rem;
            flex-direction: "column-reverse";
            .message{
                margin: 0.2rem;
                max-width: 60%;
                position: relative;
                .speech-mark {
                    position: absolute;
                    z-index: 0;
                    bottom: 1.625rem;
                    height: 20px;
                    width: 20px;
                    background: cssvar(success);
                }
                .speech-mark-negative {
                    position: absolute;
                    z-index: 1;
                    bottom: 1.625rem;
                    width: 10px;
                    height: 20px;
                    background: cssvar(light);
                }
                .message-content{
                    padding: 0.5rem 1rem;
                    border-radius: 1.25rem;
                    .image-container {
                        padding-top: 0.5rem;
                        .image {
                            min-height: 200px;
                            min-height: 100px;
                            width: 100%;
                            height: 100%;
                            background:cssvar(light);
                        }
                    }
                }
                .info-container{
                    margin: 0.2rem;
                    color: cssvar(text-color);
                    display: flex;
                    align-items: center;
                    font-size: 0.7rem;
                    white-space: nowrap;
                    yoo-avatar {
                        margin-right: 0.3rem;
                    }
                }
                &.user-message{
                    margin-left: auto;
                    .speech-mark {
                        border-bottom-left-radius: 15px;
                        right: -7px;
                    }
                    .speech-mark-negative {
                        border-bottom-left-radius: 10px;
                        right: -10px;
                    }
                    .message-content{
                        background: cssvar(success);
                        color: cssvar(light);
                    }
                    .info-container{
                        margin-left: auto;
                    }
                }
                &.other-message{
                    margin-right: auto;
                    .speech-mark {
                        left: -7px;
                        border-bottom-right-radius: 15px;
                        background: cssvar(dark-10);
                    }
                    .speech-mark-negative {
                        left: -10px;
                        border-bottom-right-radius: 10px;
                    }
                    .message-content{
                        background: cssvar(dark-10);
                    }
                }
            }
        }
        .input-container {
            position: fixed;
            bottom: 0;
            width: 100%;
            background: white;
            //padding: 0.5625rem 0.9375rem 0.4375rem 0.875rem;
            box-shadow: $input-bar-shadow;
            padding-bottom: env(safe-area-inset-bottom);
        }
    }
}