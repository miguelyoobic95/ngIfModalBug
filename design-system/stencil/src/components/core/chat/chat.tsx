import { Component, Prop, Element, Method, State, Event, EventEmitter } from '@stencil/core';
import { YooSlimScrollComponent } from '../../core/slim-scroll/slim-scroll';
import { IChatMessage } from '@shared/interfaces';
import { getBackImageStyle, cloudinary, getUserDisplayName, resizeWindow } from '../../../utils/helpers';
import { pipes } from '../../../utils/pipes';

@Component({
    tag: 'yoo-chat',
    styleUrl: 'chat.scss',
    scoped: true
})
export class YooChatComponent {

    @Prop() heading: string;
    @Prop() messages: IChatMessage[] = []; // suppose that messages are ordered to most recent to the older one
    @Prop() displayLoadMore: boolean;

    @Event() loadMoreClicked: EventEmitter<boolean>;
    @Event() sendText: EventEmitter<string>;

    @Element() host: HTMLStencilElement;

    @State() scrollHeight: string;
    @State() loaded: boolean = false;

    @Method()
    scrollToBottom(): void {
        let slim = this.host.querySelector('yoo-slim-scroll');
        if (slim) {
            if (this.loaded) {
                slim.scrollToBottom();
            } else {
                setTimeout(() => this.scrollToBottom(), 200);
            }
        }
    }

    @Method()
    scrollToTop(): void {
        let slim = this.host.querySelector('yoo-slim-scroll');
        if (slim) {
            if (this.loaded) {
                slim.scrollToTop();
            } else {
                setTimeout(() => this.scrollToTop(), 200);
            }
        }
    }

    @Method()
    resize(windowEvent: boolean = false) {
        let slim: YooSlimScrollComponent = this.host.querySelector('yoo-slim-scroll') as any;
        if (slim) {
            if (slim.height === this.getSizeContainer().height && windowEvent) {
                setTimeout(() => this.scrollToBottom(), 250);
            }
            this.scrollHeight = this.getSizeContainer().height;
        }
    }

    componentDidLoad() {
        resizeWindow(() => this.resize(true));
        this.resize();
        this.loaded = true;
    }

    componentDidUpdate() {
        this.resize();
    }

    getSizeContainer() {
        let maxHeight = window.innerHeight;
        let body = this.host.parentElement.parentElement.querySelector('.modal-body');
        if (body) {
            maxHeight = Math.min(maxHeight, body.clientHeight);
        } else {
            maxHeight = this.decreaseMaxHeight(maxHeight, 'ion-header', document);
            maxHeight = this.decreaseMaxHeight(maxHeight, 'ion-footer', document);
        }
        maxHeight = this.decreaseMaxHeight(maxHeight, '.load-more', this.host);
        maxHeight = this.decreaseMaxHeight(maxHeight, '.input-container', this.host);
        return { height: maxHeight + 'px' };
    }

    decreaseMaxHeight(maxHeight: number, name: string, html) {
        let comp: HTMLElement = html.querySelector(name);
        if (comp) {
            maxHeight -= comp.clientHeight;
        }
        return maxHeight;
    }

    getLastMessage(index: number): boolean {
        if (index === (this.messages.length - 1)) {
            return true;
        } else if (this.messages[index].author !== this.messages[index + 1].author) {
            return true;
        } else {
            return false;
        }
    }

    onSendText(ev: CustomEvent) {
        ev.stopPropagation();
        this.sendText.emit(ev.detail);
    }

    render(): JSX.Element {
        return (
            <div class="outer-container">
                <div class="chat-header">
                    <span>{this.heading}</span>
                </div>
                {this.displayLoadMore ?
                    <div class="load-more" onClick={() => this.loadMoreClicked.emit(true)} attr-layout="row">
                        <span>{'Load More'}</span>
                    </div>
                    : null}
                <div class="scroll-container">
                    <yoo-slim-scroll height={this.scrollHeight}>
                        <div class="messages-container" attr-layout="column">
                            {(this.messages.map((m, index) =>
                                <div class={'message ' + ((m.isAlternate) ? 'user-message' : 'other-message')} attr-layout="column">
                                    <div class={'message-content' + (this.getLastMessage(index) ? ' last' : '')} attr-layout="column">
                                        {m.img ?
                                            (<div class="image-container">
                                                <div class="image" style={getBackImageStyle(cloudinary(m.img))}></div> <br />
                                            </div>)
                                            : null}
                                        <span>{m.content}</span>
                                    </div>
                                    {this.getLastMessage(index) ? [
                                        <div class="speech-mark">
                                        </div>,
                                        <div class="speech-mark-negative">
                                        </div>
                                    ] : null}
                                    {(m.author || m.time) && this.getLastMessage(index) ?
                                        <div class="info-container">
                                            <yoo-avatar class="xsmall" user={m.author}></yoo-avatar>
                                            <span>{getUserDisplayName(m.author)} - {pipes.timeAgo.transform(m.time)}</span>
                                        </div>
                                        : null}
                                </div>
                            ))}
                        </div>
                    </yoo-slim-scroll>
                </div>
                <div class="input-container">
                    <yoo-input-bar onSendText={(ev) => this.onSendText(ev)}></yoo-input-bar>
                </div>
            </div>
        );
    }
}
