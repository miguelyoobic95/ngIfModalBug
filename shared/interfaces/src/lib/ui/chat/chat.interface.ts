import { IUser } from '../../entities/user/user.interface';

export interface IChatMessage {
    content: string;
    author?: IUser;
    time?: Date;
    isAlternate?: boolean;
    img?: string;
    deleted?: boolean;
}
