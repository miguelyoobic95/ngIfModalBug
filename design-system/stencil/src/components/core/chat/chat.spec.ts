import { TestWindow } from '@stencil/core/testing';
import { YooChatComponent } from './chat';

describe('YooChatComponent', () => {
    let window;
    beforeEach(() => {
        window = new TestWindow();
    });
    it('should build', () => {
        expect(new YooChatComponent()).toBeTruthy();
    });

    describe('Rendering', () => {
        it ('Should render', async () => {
            let element = await window.load({
                components: [YooChatComponent],
                html: '<yoo-chat></yoo-chat>'
            });
            expect(element).toBeTruthy();
            expect(element).toMatchSnapshot();
        });

        it ('Should render with a heading', async () => {
            let element = await window.load({
                components: [YooChatComponent],
                html: '<yoo-chat heading="Chat Heading"></yoo-chat>'
            });
            expect(element).toBeTruthy();
            expect(element).toMatchSnapshot();
        });

        it ('Should render with messages', async () => {
            let element = await window.load({
                components: [YooChatComponent],
                html: '<yoo-chat heading="Chat Heading"></yoo-chat>'
            });
            expect(element.messages).toEqual([]);
            element.messages = [{content: 'First message', author: 'Seb', time: '8:47am'},
                {content: '2nd message: It seems to work fine but I\'m sure there is a bug somewhere !',
                    author: 'Seb'},
                {content: 'Third message, it is a very very very very long message. Not so long after all !',
                    author: 'Mike', isAlternate: true}];
            await window.flush();
            expect(element).toMatchSnapshot();
        });

        it('Should render with images inside the messages', async () => {
            let element = await window.load({
                components: [YooChatComponent],
                html: '<yoo-chat heading="Chat Heading"></yoo-chat>'
            });
            expect(element.messages).toEqual([]);
            element.messages = [{content: 'First message', author: 'Seb', time: '8:47am'},
                {content: '2nd message: It seems to work fine but I\'m sure there is a bug somewhere !',
                    author: 'Seb', img: './assets/grid/empty-state-1.svg'},
                {content: 'Third message, it is a very very very very long message. Not so long after all !',
                    author: 'Mike', isAlternate: true, img: './assets/grid/empty-state-1.svg'}];
            await window.flush();
            expect(element).toMatchSnapshot();
        });
    });

    describe('Re-rendering', () => {
        it('Should re render when heading is updated', async () => {
            let element = await window.load({
                components: [YooChatComponent],
                html: '<yoo-chat heading="Chat Title"></yoo-chat>'
            });
            expect(element.heading).toEqual('Chat Title');
            element.heading = 'New Title';
            await window.flush();
            expect(element.heading).toEqual('New Title');
            expect(element).toMatchSnapshot();
        });
    });
});
