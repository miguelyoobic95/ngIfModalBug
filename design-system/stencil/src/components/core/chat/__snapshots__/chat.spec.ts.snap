// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooChatComponent Re-rendering Should re render when heading is updated 1`] = `
<yoo-chat
  class="hydrated"
  data-yoo-chat-host=""
  heading="Chat Title"
>
  
  <div
    class="outer-container"
    data-yoo-chat=""
  >
    <div
      class="chat-header"
      data-yoo-chat=""
    >
      <span
        data-yoo-chat=""
      >
        New Title
      </span>
    </div>
    
    <div
      class="scroll-container"
      data-yoo-chat=""
    >
      <yoo-slim-scroll
        data-yoo-chat=""
        height="768px"
      >
        <div
          attr-layout="column"
          class="messages-container"
          data-yoo-chat=""
        />
      </yoo-slim-scroll>
    </div>
    <div
      class="input-container"
      data-yoo-chat=""
    >
      <yoo-input-bar
        data-yoo-chat=""
      />
    </div>
  </div>
</yoo-chat>
`;

exports[`YooChatComponent Rendering Should render 1`] = `
<yoo-chat
  class="hydrated"
  data-yoo-chat-host=""
>
  
  <div
    class="outer-container"
    data-yoo-chat=""
  >
    <div
      class="chat-header"
      data-yoo-chat=""
    >
      <span
        data-yoo-chat=""
      >
        
      </span>
    </div>
    
    <div
      class="scroll-container"
      data-yoo-chat=""
    >
      <yoo-slim-scroll
        data-yoo-chat=""
        height="768px"
      >
        <div
          attr-layout="column"
          class="messages-container"
          data-yoo-chat=""
        />
      </yoo-slim-scroll>
    </div>
    <div
      class="input-container"
      data-yoo-chat=""
    >
      <yoo-input-bar
        data-yoo-chat=""
      />
    </div>
  </div>
</yoo-chat>
`;

exports[`YooChatComponent Rendering Should render with a heading 1`] = `
<yoo-chat
  class="hydrated"
  data-yoo-chat-host=""
  heading="Chat Heading"
>
  
  <div
    class="outer-container"
    data-yoo-chat=""
  >
    <div
      class="chat-header"
      data-yoo-chat=""
    >
      <span
        data-yoo-chat=""
      >
        Chat Heading
      </span>
    </div>
    
    <div
      class="scroll-container"
      data-yoo-chat=""
    >
      <yoo-slim-scroll
        data-yoo-chat=""
        height="768px"
      >
        <div
          attr-layout="column"
          class="messages-container"
          data-yoo-chat=""
        />
      </yoo-slim-scroll>
    </div>
    <div
      class="input-container"
      data-yoo-chat=""
    >
      <yoo-input-bar
        data-yoo-chat=""
      />
    </div>
  </div>
</yoo-chat>
`;

exports[`YooChatComponent Rendering Should render with images inside the messages 1`] = `
<yoo-chat
  class="hydrated"
  data-yoo-chat-host=""
  heading="Chat Heading"
>
  
  <div
    class="outer-container"
    data-yoo-chat=""
  >
    <div
      class="chat-header"
      data-yoo-chat=""
    >
      <span
        data-yoo-chat=""
      >
        Chat Heading
      </span>
    </div>
    
    <div
      class="scroll-container"
      data-yoo-chat=""
    >
      <yoo-slim-scroll
        data-yoo-chat=""
        height="768px"
      >
        <div
          attr-layout="column"
          class="messages-container"
          data-yoo-chat=""
        >
          <div
            attr-layout="column"
            class="message other-message"
            data-yoo-chat=""
          >
            <div
              attr-layout="column"
              class="message-content"
              data-yoo-chat=""
            >
              
              <span
                data-yoo-chat=""
              >
                First message
              </span>
            </div>
            
          </div>
          <div
            attr-layout="column"
            class="message other-message"
            data-yoo-chat=""
          >
            <div
              attr-layout="column"
              class="message-content last"
              data-yoo-chat=""
            >
              <div
                class="image-container"
                data-yoo-chat=""
              >
                <div
                  class="image"
                  data-yoo-chat=""
                  style="background-repeat: no-repeat; background-size: cover; background-position-x: 5;0%; background-image: url(./assets/grid/empty-state-1.svg);"
                />
                 
                <br
                  data-yoo-chat=""
                />
              </div>
              <span
                data-yoo-chat=""
              >
                2nd message: It seems to work fine but I'm sure there is a bug somewhere !
              </span>
            </div>
            <div
              class="speech-mark"
              data-yoo-chat=""
            />
            <div
              class="speech-mark-negative"
              data-yoo-chat=""
            />
            <div
              class="info-container"
              data-yoo-chat=""
            >
              <yoo-avatar
                class="xsmall"
                data-yoo-chat=""
                user="Seb"
              />
              <span
                data-yoo-chat=""
              >
                 - 
              </span>
            </div>
          </div>
          <div
            attr-layout="column"
            class="message user-message"
            data-yoo-chat=""
          >
            <div
              attr-layout="column"
              class="message-content last"
              data-yoo-chat=""
            >
              <div
                class="image-container"
                data-yoo-chat=""
              >
                <div
                  class="image"
                  data-yoo-chat=""
                  style="background-repeat: no-repeat; background-size: cover; background-position-x: 5;0%; background-image: url(./assets/grid/empty-state-1.svg);"
                />
                 
                <br
                  data-yoo-chat=""
                />
              </div>
              <span
                data-yoo-chat=""
              >
                Third message, it is a very very very very long message. Not so long after all !
              </span>
            </div>
            <div
              class="speech-mark"
              data-yoo-chat=""
            />
            <div
              class="speech-mark-negative"
              data-yoo-chat=""
            />
            <div
              class="info-container"
              data-yoo-chat=""
            >
              <yoo-avatar
                class="xsmall"
                data-yoo-chat=""
                user="Mike"
              />
              <span
                data-yoo-chat=""
              >
                 - 
              </span>
            </div>
          </div>
        </div>
      </yoo-slim-scroll>
    </div>
    <div
      class="input-container"
      data-yoo-chat=""
    >
      <yoo-input-bar
        data-yoo-chat=""
      />
    </div>
  </div>
</yoo-chat>
`;

exports[`YooChatComponent Rendering Should render with messages 1`] = `
<yoo-chat
  class="hydrated"
  data-yoo-chat-host=""
  heading="Chat Heading"
>
  
  <div
    class="outer-container"
    data-yoo-chat=""
  >
    <div
      class="chat-header"
      data-yoo-chat=""
    >
      <span
        data-yoo-chat=""
      >
        Chat Heading
      </span>
    </div>
    
    <div
      class="scroll-container"
      data-yoo-chat=""
    >
      <yoo-slim-scroll
        data-yoo-chat=""
        height="768px"
      >
        <div
          attr-layout="column"
          class="messages-container"
          data-yoo-chat=""
        >
          <div
            attr-layout="column"
            class="message other-message"
            data-yoo-chat=""
          >
            <div
              attr-layout="column"
              class="message-content"
              data-yoo-chat=""
            >
              
              <span
                data-yoo-chat=""
              >
                First message
              </span>
            </div>
            
          </div>
          <div
            attr-layout="column"
            class="message other-message"
            data-yoo-chat=""
          >
            <div
              attr-layout="column"
              class="message-content last"
              data-yoo-chat=""
            >
              
              <span
                data-yoo-chat=""
              >
                2nd message: It seems to work fine but I'm sure there is a bug somewhere !
              </span>
            </div>
            <div
              class="speech-mark"
              data-yoo-chat=""
            />
            <div
              class="speech-mark-negative"
              data-yoo-chat=""
            />
            <div
              class="info-container"
              data-yoo-chat=""
            >
              <yoo-avatar
                class="xsmall"
                data-yoo-chat=""
                user="Seb"
              />
              <span
                data-yoo-chat=""
              >
                 - 
              </span>
            </div>
          </div>
          <div
            attr-layout="column"
            class="message user-message"
            data-yoo-chat=""
          >
            <div
              attr-layout="column"
              class="message-content last"
              data-yoo-chat=""
            >
              
              <span
                data-yoo-chat=""
              >
                Third message, it is a very very very very long message. Not so long after all !
              </span>
            </div>
            <div
              class="speech-mark"
              data-yoo-chat=""
            />
            <div
              class="speech-mark-negative"
              data-yoo-chat=""
            />
            <div
              class="info-container"
              data-yoo-chat=""
            >
              <yoo-avatar
                class="xsmall"
                data-yoo-chat=""
                user="Mike"
              />
              <span
                data-yoo-chat=""
              >
                 - 
              </span>
            </div>
          </div>
        </div>
      </yoo-slim-scroll>
    </div>
    <div
      class="input-container"
      data-yoo-chat=""
    >
      <yoo-input-bar
        data-yoo-chat=""
      />
    </div>
  </div>
</yoo-chat>
`;
